contracts/native-staking-proxy/src/contract.rs

```diff
@@ -0,0 +1,167 @@
+use cosmwasm_std::{ensure_eq, DistributionMsg, Response, Uint128, VoteOption, WeightedVoteOption};
+use cw2::set_contract_version;
+use cw_storage_plus::Item;
+
+use sylvia::types::{ExecCtx, InstantiateCtx, QueryCtx};
+use sylvia::{contract, schemars};
+
+use crate::error::ContractError;
+use crate::types::{ClaimsResponse, Config, ConfigResponse};
+
+pub const CONTRACT_NAME: &str = env!("CARGO_PKG_NAME");
+pub const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+pub struct NativeStakingProxyContract<'a> {
+    // TODO
+    config: Item<'a, Config>,
+}
+
+#[contract(error=ContractError)]
+impl NativeStakingProxyContract<'_> {
+    pub const fn new() -> Self {
+        Self {
+            config: Item::new("config"),
+        }
+    }
+
+    /// The caller of the instantiation will be the native-staking contract.
+    /// We stake `funds.info` on the given validator
+    #[msg(instantiate)]
+    pub fn instantiate(
+        &self,
+        ctx: InstantiateCtx,
+        denom: String,
+        owner: String,
+        validator: String,
+    ) -> Result<Response, ContractError> {
+        let config = Config {
+            denom,
+            parent: ctx.info.sender.clone(),
+            owner: ctx.deps.api.addr_validate(&owner)?,
+        };
+        self.config.save(ctx.deps.storage, &config)?;
+        set_contract_version(ctx.deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+        // stake info.funds on validator
+        let res = self.stake(ctx, validator)?;
+
+        // set owner as recipient of future withdrawls
+        let set_withdrawl = DistributionMsg::SetWithdrawAddress {
+            address: config.parent.into_string(),
```

Comment:
```suggestion
            address: config.owner.into_string(),
```

?

If I understand this properly:

 - `sender` would (typically) be the vault contract, and becomes `parent` here.
 - `owner` would be (typically) set by the vault to the original user, the provider of the funds.

In principle it should be possible to instantiate one of these contracts directly, and that is not a problem. The vault would be just a convenient way to do it, AFAIK.

Reply:
You are right.

Reply:
Yes, typo. And good catch (why we need tests and real implementation). I just started sketching it out